orchestrator/lib/az/adb2c/tokens.go

```diff
@@ -0,0 +1,533 @@
+		}
+
+		// If empty, use the expected issuer
+		if config.Issuer == "" {
+			config.Issuer = issuer
+		}
+
+		c.issuerConfigs[issuer] = config
+
+		if err := c.jwksFetcher.Register(config.JwksURI); err != nil {
+			return fmt.Errorf("failed to register JWKS URI: %w", err)
+		}
+
+		// Force a refresh of the JWKS to ensure it's loaded and processed
+		keySet, err := c.jwksFetcher.Refresh(ctx, config.JwksURI)
+		if err != nil {
+			return fmt.Errorf("failed to refresh JWKS from %s: %w", config.JwksURI, err)
+		}
+
+		c.ensureKeyAlgorithms(ctx, keySet)
+	}
+
+	c.lastRefresh = time.Now()
+	return nil
+}
+
+// fetchOpenIDConfigurationFromURL fetches the OpenID configuration from a specific URL
+func (c *Client) fetchOpenIDConfigurationFromURL(ctx context.Context, discoveryURL string) (*oidc.DiscoveryConfiguration, error) {
+	// Create a context with timeout for security
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, discoveryURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %w", err)
+	}
+
+	// Add security headers
+	req.Header.Set("User-Agent", "ORCA-ADB2C-Client/1.0")
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Cache-Control", "no-cache")
+
+	// Use a secure HTTP client with timeouts
+	client := &http.Client{
+		Timeout: 30 * time.Second,
+		Transport: &http.Transport{
+			TLSHandshakeTimeout:   10 * time.Second,
+			ResponseHeaderTimeout: 10 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+		},
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch OpenID configuration: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed to fetch OpenID configuration: status code %d", resp.StatusCode)
+	}
+
+	// Limit response size to prevent DoS attacks
+	limitedReader := &io.LimitedReader{R: resp.Body, N: 1024 * 1024} // 1MB limit
+
+	var config oidc.DiscoveryConfiguration
+	if err := json.NewDecoder(limitedReader).Decode(&config); err != nil {
+		return nil, fmt.Errorf("failed to parse OpenID configuration: %w", err)
+	}
+
+	return &config, nil
+}
+
+// fetchKeySet fetches the JWK set for a given issuer, refreshing if needed
+func (c *Client) fetchKeySet(ctx context.Context, issuer string) (jwk.Set, error) {
+	c.issuerConfigsMutex.RLock()
+	issuerConfig, ok := c.issuerConfigs[issuer]
+	c.issuerConfigsMutex.RUnlock()
+
+	if !ok {
+		return nil, fmt.Errorf("no configuration found for issuer: %s", issuer)
+	}
+
+	// Check if refresh is needed
+	if time.Since(c.lastRefresh) >= c.refreshInterval {
+		// Since we do not store the last refresh time per issuer/token, we refresh all configurations
+		if err := c.refreshAllConfigurations(ctx); err != nil {
+			return nil, fmt.Errorf("failed to refresh configurations: %w", err)
+		}
```

Comment:
do we expect to need multiple trusted issuers? Since we're adding risk here (one openid configuraiton expired, need to refresh all, the second one is down blocking all access). I'd say we support it properly, or just support 1?